src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1405,26 +1407,111 @@ private Type EmitFunctionCallNode(AssociativeNode node)
-                int index = 0;
-                foreach(var arg in args)
+                //TODO if possible only marshal args that resulted from a replicated call.
+                //IF the arguments of this function call were returned from a replicated call -
+                //we need to unmarshal them from CLRStackValue wrappers to plain # objects -
+                // to do this, we need use the methodcache to get the methodinfo
+                // where we can then get the paraminfo we'll use for unmarshling.
+                //
+
+                
+                // Emit methodCache passed as arg to global Execute method.
+                EmitOpCode(OpCodes.Ldarg_1);
+                // Emit className for input to call to CodeGenIL.KeyGen
+                EmitOpCode(OpCodes.Ldstr, className);
+
+                // Emit methodName for input to call to CodeGenIL.KeyGen
+                EmitOpCode(OpCodes.Ldstr, methodName);
+                EmitOpCode(OpCodes.Ldc_I4, numArgs);
+                var keygen = typeof(CodeGenIL).GetMethod(nameof(CodeGenIL.KeyGen));
+                EmitOpCode(OpCodes.Call, keygen);
+
+                var local = DeclareLocal(typeof(IEnumerable<ProtoCore.CLRFunctionEndPoint>), "cached MethodBase objects");
+                //local could be null if we are not emitting currently.
+                if (local != null)
                 {
-                    Type t = DfsTraverse(arg);
+                    EmitOpCode(OpCodes.Ldloca, local);
+                }
+
+                // Emit methodCache.TryGetValue(KeyGen(...), out IEnumerable<CLRFunctionEndPoint> feps)
+                var dictLookup = typeof(IDictionary<int, IEnumerable<ProtoCore.CLRFunctionEndPoint>>).GetMethod(
+                    nameof(IDictionary<int, IEnumerable<ProtoCore.CLRFunctionEndPoint>>.TryGetValue));
+                EmitOpCode(OpCodes.Callvirt, dictLookup);
+
+                EmitOpCode(OpCodes.Pop);
+                if (local != null)
+                {
+                    EmitOpCode(OpCodes.Ldloc, local.LocalIndex);
+                }
+
+                //emit an array of args to pass to unmarshal
+                EmitArray(typeof(object), args, (AssociativeNode n, int index) =>
+                {
+                    Type t = DfsTraverse(n);
                     if (isStaticOrCtor)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index]);
+                        t = EmitCoercionCode(n, t, parameters[index]);
                     }
                     else if (index > 0)
                     {
-                        t = EmitCoercionCode(arg, t, parameters[index - 1]);
+                        t = EmitCoercionCode(n, t, parameters[index - 1]);
+                    }
+
+                    if (t == null) return;
+
+                    if (t.IsValueType)
+                    {
+                        EmitOpCode(OpCodes.Box, t);
+                    }
+                });
```

Comment:
Why does all this need to be done instead of simply `Type t = DfsTraverse(n);` and `EmitCoercion` in a for loop? Just trying to figure out what was wrong in the previous logic.

Reply:
Oh, I see this is to emit the args for `UnMarshal2`. It looks like all this is negating the benefits of a direction function call 🤔

Reply:
well - thats not exactly clear yet - we **need** to do the un-marshaling somewhere between a replicated call and a non replicated one.

I have yet to get some hard numbers.

It definitely ends up making the code more complex - please see point 5 in the PR description for more info.